// @nguniversal/common/fesm5/common.js
import { HTTP_INTERCEPTORS, HttpHeaders, HttpResponse } from '@angular/common/http';
import { APP_INITIALIZER, ApplicationRef, Injectable, NgModule } from '@angular/core';
import { BrowserTransferStateModule, TransferState, makeStateKey } from '@angular/platform-browser';
import { of } from 'rxjs';
import { filter, take, tap } from 'rxjs/operators';
import { DOCUMENT } from '@angular/common';

var __values = (undefined && undefined.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
function getHeadersMap(headers) {
    var headersMap = {};
    try {
        for (var _a = __values(headers.keys()), _b = _a.next(); !_b.done; _b = _a.next()) {
            var key = _b.value;
            headersMap[key] = (headers.getAll(key));
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return headersMap;
    var e_1, _c;
}
var TransferHttpCacheInterceptor = /** @class */ (function () {
    function TransferHttpCacheInterceptor(appRef, transferState) {
        var _this = this;
        this.transferState = transferState;
        this.isCacheActive = true;
        // Stop using the cache if the application has stabilized, indicating initial rendering is
        // complete.
        appRef.isStable
            .pipe(filter(function (isStable) { return isStable; }), take(1)).toPromise()
            .then(function () { _this.isCacheActive = false; });
    }
    TransferHttpCacheInterceptor.prototype.invalidateCacheEntry = function (url) {
        this.transferState.remove(makeStateKey('G.' + url));
        this.transferState.remove(makeStateKey('H.' + url));
    };
    TransferHttpCacheInterceptor.prototype.intercept = function (req, next) {
        var _this = this;
        // Stop using the cache if there is a mutating call.
        if (req.method !== 'GET' && req.method !== 'HEAD') {
            this.isCacheActive = false;
            this.invalidateCacheEntry(req.url);
        }
        if (!this.isCacheActive) {
            // Cache is no longer active. Pass the request through.
            return next.handle(req);
        }
        var key = (req.method === 'GET' ? 'G.' : 'H.') + req.url;
        var storeKey = makeStateKey(key);
        if (this.transferState.hasKey(storeKey)) {
            // Request found in cache. Respond using it.
            var response = this.transferState.get(storeKey, {});
            return of(new HttpResponse({
                body: response.body,
                headers: new HttpHeaders(response.headers),
                status: response.status,
                statusText: response.statusText,
                url: response.url,
            }));
        }
        else {
            // Request not found in cache. Make the request and cache it.
            var httpEvent = next.handle(req);
            return httpEvent
                .pipe(tap(function (event) {
                if (event instanceof HttpResponse) {
                    _this.transferState.set(storeKey, {
                        body: event.body,
                        headers: getHeadersMap(event.headers),
                        status: event.status,
                        statusText: event.statusText,
                        url: (event.url),
                    });
                }
            }));
        }
    };
    TransferHttpCacheInterceptor.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TransferHttpCacheInterceptor.ctorParameters = function () { return [
        { type: ApplicationRef, },
        { type: TransferState, },
    ]; };
    return TransferHttpCacheInterceptor;
}());
/**
 * An NgModule used in conjunction with `ServerTransferHttpCacheModule` to transfer cached HTTP
 * calls from the server to the client application.
 */
var TransferHttpCacheModule = /** @class */ (function () {
    function TransferHttpCacheModule() {
    }
    TransferHttpCacheModule.decorators = [
        { type: NgModule, args: [{
                    imports: [BrowserTransferStateModule],
                    providers: [
                        TransferHttpCacheInterceptor,
                        { provide: HTTP_INTERCEPTORS, useExisting: TransferHttpCacheInterceptor, multi: true },
                    ],
                },] }
    ];
    return TransferHttpCacheModule;
}());

/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
function domContentLoadedFactory(doc) {
    return function () {
        return new Promise(function (resolve, _reject) {
            var contentLoaded = function () {
                doc.removeEventListener('DOMContentLoaded', contentLoaded);
                resolve();
            };
            doc.addEventListener('DOMContentLoaded', contentLoaded);
        });
    };
}
var StateTransferInitializerModule = /** @class */ (function () {
    function StateTransferInitializerModule() {
    }
    StateTransferInitializerModule.decorators = [
        { type: NgModule, args: [{
                    providers: [
                        { provide: APP_INITIALIZER, multi: true, useFactory: domContentLoadedFactory, deps: [DOCUMENT] },
                    ]
                },] }
    ];
    return StateTransferInitializerModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { domContentLoadedFactory as ɵnguniversal_modules_common_common_b, TransferHttpCacheInterceptor as ɵnguniversal_modules_common_common_a, TransferHttpCacheModule, StateTransferInitializerModule };
